src/Message/UniformMPIError.h

```diff
@@ -0,0 +1,42 @@
+//////////////////////////////////////////////////////////////////////////////////////
+// This file is distributed under the University of Illinois/NCSA Open Source License.
+// See LICENSE file in top directory for details.
+//
+// Copyright (c) 2020 QMCPACK developers.
+//
+// File developed by: Peter Doak, doakpw@ornl.gov, Oak Ridge National Lab
+//
+// File created by: Peter Doak, doakpw@ornl.gov, Oak Ridge National Lab
+//////////////////////////////////////////////////////////////////////////////////////
+#ifndef QMCPLUSPLUS_UNIVERSALERROR_HPP
+#define QMCPLUSPLUS_UNIVERSALERROR_HPP
+
+#include <exception>
+namespace qmcplusplus
+{
+/** This a subclass for runtime errors that will occur on all ranks.
+ *
+ *  This combined with a block like which must call barrier_and_abort for the 
+ *  MPI comm the error is thrown in.
+ * 
+ *  catch (const UniformMPIerror& ue)
+ *  {
+ *    my_local_comm->barrier_and_abort(ue.what());
+ *  }
+ *
+ *  insures that the error message actually makes it out on the head
+ *  rank before the MPI processes are aborted.
+ *  
+ *  If even one rank does not experience the issue you may hang the
+ *  job until the walltime expires.
+ */
+class UniformMPIError : public std::runtime_error
```

Comment:
I'm not really excited about the name, since it describes implementation details.     More precisely, it's the use of this class without creating a better named derived class.     The name makes me think there is something wrong related to MPI.

Reply:
The problem here is that C++ and MPI have their own distinct error mechanisms.

I suggest UniformError for the name. When MPI is not available we can still safely call them.

Comments could be expanded to indicate that these functions are the preferred way to give errors since they are exception & MPI aware.

Reply:
`UniformErrorForTopCommunicatorInCallStack`?

Reply:
I think of the call stack going down like most debuggers I've used, so that name would be confusing to me. 

I think `UniformErrorInCommunicateScope` could work. Since Mark is right It does match the abstraction level of Communicate. Which I didn't initally notice since I have the bad habit of replacing Communicate with MPI mentally. 

This issue doesn't play out for thead level concurrency the same so I don't think `UniformError` is a good idea. For the thread level we already need to catch any exception thrown in a section run by a `ParallelExecutor` (PE) (at least for the std::thread and openmp implementations) similarly. Since if you don't catch the exception in the thread scope both of those "runtimes" call terminate, and then the rank crashes without calling MPI_Abort (I think). There is an implicit barrier at the close of the PE::operator(...) and after that barrier is passed an exception is thrown at the enclosing thread scope (generally the master thread) if any of the threads encountered and exception. That exception is caught by a catch that calls comm->abort(). 

We could easily write a slightly more complicated implementation for PE to handle uniform errors as a special case since we could pass out the exception information for every thread. This removes the burden from the developer of decided whether an exception would be uniform or not at the thread scope everywhere. At the communicate level without more complex comm code (more than the 10 lines or so it would be for PE) we don't have the information about other ranks exceptions, but the developer might have a good idea. The user experience is improved by seeing the error message everytime not just when rank 0 won the race.

I realize we want to treat the UniformErrorInCommunicateScope differently if thrown inside a PE, right now they would not result in same exception being thrown in the enclosing thread scope which is bad if you decide to say use paralellism parsing input.

Reply:
We should not parsing input in any PE scope. The parsing of non-local move input in the batched driver results tons of printing messages. Even if we can stop printing on non-zero thread but I think the right fix is to parse the input outside PE and than handle an intermediate structure to inside PE. So we don't need to worry about inside PE case, we just don't support that trouble case at all.